File: 프로그래머스/3/43162. 네트워크/네트워크.py
import sys
from collections import deque
input = sys.stdin.readline

def bfs(node, graph, visited):
    q = deque([node])
    
    while q:
        node = q.popleft()
        
        for i in graph[node]:
            if not visited[i]:
                visited[i] = True
                q.append(i)
                
    return 1

def solution(n, computers):
    answer = 0
    graph = [[] for _ in range(n+1)]
    visited = [False] * (n+1)
    
    for i in range(n):
        for j in range(n):
            if computers[i][j]:
                graph[i+1].append(j+1)
                
    for i in range(1, n+1):
        if not visited[i]:
            visited[i] = True
            answer += bfs(i, graph, visited)
    
    print(graph)
    return answer

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int bfs(int node, vector<int> &v, vector<vector<int>> &graph) {
    queue<int> q;
    q.push(node);
    
    while (!q.empty()) {
        int now_node = q.front();
        q.pop();
        
        for (int i = 0; i < graph[now_node].size(); i++) {
            if (v[graph[now_node][i]] == 1) continue;
        
            v[graph[now_node][i]] = 1;
            q.push(graph[now_node][i]);
        }
    }
    
    return 1;
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    vector<int> v(n);
    vector<vector<int>> graph(n);
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (i == j || computers[i][j] == 0) continue;
            
            graph[i].push_back(j);
        }
    }
    
    for (int i = 0; i < n; i++) {
        if (v[i] == 1) continue;
        
        v[i] = 1;
        answer += bfs(i, v, graph);
    }
    
    return answer;
}